ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs

```diff
@@ -50,6 +55,89 @@ data DomainAddress = DomainAddress {
   deriving (Show, Eq, Ord)
 
 
+-----------------------------------------------
+-- Resource
+--
+
+-- | Evolving resource; We use it to reinitialise the dns library if the
+-- `/etc/resolv.conf` file was modified.
+--
+data Resource a = Resource {
+    withResource :: IO (a, Resource a)
+  }
+
+constantResource :: a -> Resource a
+constantResource a = Resource (pure (a, constantResource a))
+
+-- |
+--
+-- TODO: it could be useful for `publicRootPeersProvider`.
+--
+_resolverResource :: DNS.ResolvConf -> IO (Resource DNS.Resolver)
+_resolverResource resolvConf = do
+    rs <- DNS.makeResolvSeed resolvConf
+    case DNS.resolvInfo resolvConf of
+      DNS.RCFilePath filePath ->
+        DNS.withResolver rs $ \resolver -> pure $ go filePath resolver Nothing
+
+      _ -> DNS.withResolver rs (pure . constantResource)
+
+  where
+    go :: FilePath -> DNS.Resolver -> Maybe UTCTime -> Resource DNS.Resolver
+    go filePath resolver Nothing = Resource $ do
+      modTime <- getModificationTime filePath
+      pure (resolver, go filePath resolver (Just modTime))
+    go filePath resolver (Just modTime) = Resource $ do
+      modTime' <- getModificationTime filePath
+      if modTime' <= modTime
+        then pure (resolver, go filePath resolver (Just modTime))
+        else do
+          rs <- DNS.makeResolvSeed resolvConf
+          DNS.withResolver rs $ \resolver' ->
+            pure (resolver', go filePath resolver (Just modTime'))
+
+
+-- | Strict version of 'Maybe' adjusted to the needs ot 'asyncResolverResource'.
+--
+data TimedResolver
+    = TimedResolver !DNS.Resolver !UTCTime
+    | NoResolver
+
+
+-- | `Resource` which passes the 'DNS.Resolver' through a 'TVar'.  Better than
+-- 'resolverResource' when using in multiple threads.
+--
+asyncResolverResource :: DNS.ResolvConf -> IO (Resource DNS.Resolver)
+asyncResolverResource resolvConf =
+    case DNS.resolvInfo resolvConf of
+      DNS.RCFilePath filePath -> do
+        resourceVar <- newTVarM NoResolver
+        pure $ go filePath resourceVar
+      _ -> do
+        rs <- DNS.makeResolvSeed resolvConf
+        DNS.withResolver rs (pure . constantResource)
+  where
+    go :: FilePath -> TVar IO TimedResolver -> Resource DNS.Resolver
+    go filePath resourceVar = Resource $ do
+      r <- atomically (readTVar resourceVar)
+      case r of
+        NoResolver -> do
+          !modTime <- getModificationTime filePath
```

Comment:
`getModificationTime` can throw `isDoesNotExistError` if the resolvConf file is missing, which will happen on OSX during temporary network failures. How will your code deal with that?

Reply:
I pushed a fix which catches `IOException`s and `DNSError`s.  Detection of network configuration changes would not workk on Windows, where the network config is not read from a file, but returned by a win32 api call.  I tested this on Linux and Windows. Check the updated branch.